g_python/htools.py

```diff
@@ -33,34 +34,90 @@ def __init__(self, ext: Extension, room_users='Users', room_model='RoomReady', r
 
         self.room_users = {}
         self.__callback_new_users = None
+        self.__callback_remove_user = None
 
         self.__ext = ext
         self.__request_id = request
+        self.__lock = threading.Lock()
+        self.lock = self.__lock
 
         ext.intercept(Direction.TO_CLIENT, self.__load_room_users, room_users)
         ext.intercept(Direction.TO_CLIENT, self.__clear_room_users, room_model)  # (clear users / new room entered)
         ext.intercept(Direction.TO_CLIENT, self.__remove_user, remove_user)
+        ext.intercept(Direction.TO_CLIENT, self.__on_status, status)
+
 
     def __remove_user(self, message: HMessage):
-        user = message.packet.read_string()
-        index = int(user)
-        if index in self.room_users:
-            del self.room_users[index]
+        self.__start_remove_user_processing_thread(int(message.packet.read_string()))
+
+    def __start_remove_user_processing_thread(self, index: int):
+        thread = threading.Thread(target=self.__process_remove_user, args=(index,))
+        thread.start()
+
+    def __process_remove_user(self, index: int):
+        self.__lock.acquire()
+        try:
+            if index in self.room_users:
+                if self.__callback_remove_user is not None:
+                    self.__callback_remove_user(self.room_users[index])
+                del self.room_users[index]
+        finally:
+            self.__lock.release()
 
     def __load_room_users(self, message: HMessage):
         users = HEntity.parse(message.packet)
-        for user in users:
-            self.room_users[user.index] = user
+        self.__start_user_processing_thread(users)
```

Comment:
this being a thread will result in the callback happening before the updates are applied

Reply:
I basically just copied it from the Unity parsers, and didn't notice that! I will try to fix it later today or tomorrow